include/pybind11/pybind11.h

```diff
@@ -885,11 +885,21 @@ class class_ : public detail::generic_type {
 
     template <typename... Extra>
     class_(handle scope, const char *name, const Extra &... extra) {
-        detail::type_record record;
+        using namespace detail;
+
+        // MI can only be specified via class_ template options, not constructor parameters
+        static_assert(
+            none_of<is_pyobject<Extra>...>::value || // no base class arguments, or:
+            (   constexpr_sum(is_pyobject<Extra>::value...) == 1 && // Exactly one base
+                constexpr_sum(is_base<options>::value...)   == 0 && // no template option bases
+                none_of<std::is_same<multiple_inheritance, Extra>...>::value), // no multiple_inheritance attr
```

Comment:
It's pretty awesome that this can be a compile-time error!

One minor point, and feel free to ignore this since it's more about code style. The first `none_of` and the first `constexpr_sum` are repeating a bit of work. Consider assigning the sum to a variable to reuse the result. In general, assigning to variables would make the code more self-documenting, avoiding the need for any extra comments. E.g.:

```c++
constexpr auto compiletime_bases = constexpr_sum(is_pyobject<options>::value...);
constexpr auto runtime_bases     = constexpr_sum(is_pyobject<Extra>::value...);
constexpr auto needs_multiple_inheritance = compiletime_bases + runtime_bases > 1 ||
                                            any_of<std::is_same<multiple_inheritance, Extra>...>();
static_assert(!(needs_multiple_inheritance && runtime_bases > 0),
              "Error: multiple inheritance bases must be specified via class_ template options");
```

(In addition to the variable names, the error message itself shows the overall intent.)